File: backend/app/services/scraper/sources/getmatch_filters.py
"""GetMatch listing filters — URL shape from getmatch.ru/vacancies."""

from __future__ import annotations

from urllib.parse import urlencode

GETMATCH_ORIGIN = "https://getmatch.ru"

SPECIALTIES = [
    ("python", "Python"),
    ("golang", "Go"),
    ("java_scala", "Java / Scala"),
    ("js_frontend", "JS / TS"),
    ("js_backend", "Node.js"),
    ("fullstack", "Fullstack"),
    ("qa_auto", "QA Auto"),
    ("qa_manual", "QA Manual"),
    ("dev_ops", "DevOps"),
    ("data_science", "ML / DS"),
    ("android", "Android"),
    ("ios", "iOS"),
    ("c_sharp", "C#"),
    ("php", "PHP"),
    ("kotlin", "Kotlin"),
    ("system_analyst", "Системный аналитик"),
    ("product_management", "Product"),
]

LOCATIONS = [
    ("remote", "удалённо"),
    ("moscow", "Москва"),
    ("saint_petersburg", "Санкт-Петербург"),
]

LEVELS = [
    ("junior", "junior"),
    ("middle", "middle"),
    ("senior", "senior"),
    ("lead", "lead"),
    ("c-level", "c-level"),
]

SALARIES = [
    (0, "любая"),
    (150000, "от 150 000"),
    (200000, "от 200 000"),
    (250000, "от 250 000"),
    (350000, "от 350 000"),
]


def _as_list(value: object) -> list[str]:
    if value is None:
        return []
    if isinstance(value, str):
        return [value] if value else []
    if isinstance(value, list):
        return [str(item) for item in value if item]
    return []


def normalize_getmatch_params(raw: dict | None) -> dict:
    data = dict(raw or {})
    allowed_sp = {item[0] for item in SPECIALTIES}
    allowed_loc = {item[0] for item in LOCATIONS}
    allowed_lvl = {item[0] for item in LEVELS}
    specialties = [
        item
        for item in _as_list(data.get("specialties") or data.get("specialty") or data.get("sp"))
        if item in allowed_sp
    ]
    unique_sp: list[str] = []
    seen_sp: set[str] = set()
    for item in specialties:
        if item not in seen_sp:
            seen_sp.add(item)
            unique_sp.append(item)
    specialties = unique_sp
    specialty = specialties[0] if specialties else ""
    location = str(data.get("location") or data.get("l") or "").strip()
    if location not in allowed_loc:
        location = ""
    levels = [item for item in _as_list(data.get("level") or data.get("g")) if item in allowed_lvl]
    salary = data.get("salary") or data.get("salary_from") or 0
    try:
        salary_n = int(salary)
    except (TypeError, ValueError):
        salary_n = 0
    return {
        "search": (data.get("search") or data.get("q") or "").strip(),
        "specialty": specialty,
        "specialties": specialties,
        "location": location,
        "level": levels,
        "salary": salary_n if salary_n > 0 else 0,
    }


def listing_url_from_params(params: dict, *, specialty: str | None = None) -> str:
    data = normalize_getmatch_params(params)
    query: list[tuple[str, str]] = []
    sp = specialty if specialty is not None else data["specialty"]
    if sp:
        query.append(("sp", sp))
    if data["location"]:
        query.append(("l", data["location"]))
    for level in data["level"]:
        query.append(("g", level))
    if data["salary"]:
        query.append(("salary_from", str(data["salary"])))
    if data["search"]:
        query.append(("q", data["search"]))
    encoded = urlencode(query, doseq=True)
    path = f"{GETMATCH_ORIGIN}/vacancies"
    return f"{path}?{encoded}" if encoded else path


def auto_name(params: dict) -> str:
    data = normalize_getmatch_params(params)
    labels = {
        **dict(SPECIALTIES),
        **dict(LOCATIONS),
    }
    parts: list[str] = []
    if data["search"]:
        parts.append(data["search"])
    specialties = data.get("specialties") or ([data["specialty"]] if data["specialty"] else [])
    if len(specialties) >= 8:
        parts.append("все специальности")
    elif len(specialties) > 1:
        parts.append(f"{len(specialties)} специальностей")
    elif specialties:
        parts.append(labels.get(specialties[0], specialties[0]))
    if data["location"]:
        parts.append(labels.get(data["location"], data["location"]))
    parts.extend(data["level"])
    if data["salary"]:
        parts.append(f"от {data['salary']}")
    unique = list(dict.fromkeys(parts))
    return " · ".join(unique[:5]) or "GetMatch поиск"

File: frontend/lib/getmatch-filters.ts
export const GETMATCH_SPECIALTIES = [
  { value: "python", label: "Python" },
  { value: "golang", label: "Go" },
  { value: "java_scala", label: "Java / Scala" },
  { value: "js_frontend", label: "JS / TS" },
  { value: "js_backend", label: "Node.js" },
  { value: "fullstack", label: "Fullstack" },
  { value: "qa_auto", label: "QA Auto" },
  { value: "qa_manual", label: "QA Manual" },
  { value: "dev_ops", label: "DevOps" },
  { value: "data_science", label: "ML / DS" },
  { value: "android", label: "Android" },
  { value: "ios", label: "iOS" },
  { value: "c_sharp", label: "C#" },
  { value: "php", label: "PHP" },
  { value: "kotlin", label: "Kotlin" },
  { value: "system_analyst", label: "Системный аналитик" },
  { value: "product_management", label: "Product" },
] as const;

export const GETMATCH_LOCATIONS = [
  { value: "remote", label: "удалённо" },
  { value: "moscow", label: "Москва" },
  { value: "saint_petersburg", label: "Санкт-Петербург" },
] as const;

export const GETMATCH_LEVELS = [
  { value: "junior", label: "junior" },
  { value: "middle", label: "middle" },
  { value: "senior", label: "senior" },
  { value: "lead", label: "lead" },
  { value: "c-level", label: "c-level" },
] as const;

export const GETMATCH_SALARIES = [
  { value: "0", label: "любая" },
  { value: "150000", label: "от 150 000" },
  { value: "200000", label: "от 200 000" },
  { value: "250000", label: "от 250 000" },
  { value: "350000", label: "от 350 000" },
] as const;

export type GetMatchFilters = {
  search: string;
  specialty: string;
  specialties?: string[];
  location: string;
  level: string[];
  salary: number;
};

export const EMPTY_GETMATCH_FILTERS: GetMatchFilters = {
  search: "",
  specialty: "",
  specialties: [],
  location: "",
  level: [],
  salary: 0,
};

function asList(value: unknown): string[] {
  if (Array.isArray(value)) return value.map(String).filter(Boolean);
  if (typeof value === "string" && value) return [value];
  return [];
}

export function getmatchFiltersFromParams(params: Record<string, unknown> | undefined): GetMatchFilters {
  const data = params ?? {};
  const salary = Number(data.salary || data.salary_from || 0);
  const specialties = asList(data.specialties || data.specialty || data.sp);
  return {
    search: String(data.search || data.q || ""),
    specialty: specialties[0] || String(data.specialty || data.sp || ""),
    specialties,
    location: String(data.location || data.l || ""),
    level: asList(data.level || data.g),
    salary: Number.isFinite(salary) && salary > 0 ? salary : 0,
  };
}

const LABELS: Record<string, string> = Object.fromEntries([
  ...GETMATCH_SPECIALTIES,
  ...GETMATCH_LOCATIONS,
].map((item) => [item.value, item.label]));

export function getmatchAutoName(filters: GetMatchFilters): string {
  const parts: string[] = [];
  if (filters.search.trim()) parts.push(filters.search.trim());
  const specialties = filters.specialties?.length ? filters.specialties : filters.specialty ? [filters.specialty] : [];
  if (specialties.length >= 8) parts.push("все специальности");
  else if (specialties.length > 1) parts.push(`${specialties.length} специальностей`);
  else if (specialties[0]) parts.push(LABELS[specialties[0]] ?? specialties[0]);
  if (filters.location) parts.push(LABELS[filters.location] ?? filters.location);
  parts.push(...filters.level);
  if (filters.salary) parts.push(`от ${filters.salary}`);
  return [...new Set(parts)].slice(0, 5).join(" · ") || "GetMatch поиск";
}

export function getmatchSummarize(filters: GetMatchFilters): string {
  const bits = [
    filters.search.trim(),
    filters.specialty ? (LABELS[filters.specialty] ?? filters.specialty) : "",
    filters.location ? (LABELS[filters.location] ?? filters.location) : "",
    ...filters.level,
    filters.salary ? `от ${filters.salary}` : "",
  ].filter(Boolean);
  return [...new Set(bits)].join(" · ") || "без фильтров";
}
